// clipper-plugin-weixin/dist/index.js
import { JSDOM } from 'jsdom'
import TurndownService from 'turndown'

function parseWeixinArticle(html = '') {
  const dom = new JSDOM(html)
  const document = dom.window.document
  const jsContent = document.getElementById('js_content')

  if (!jsContent) {
    return null
  }

  document.querySelectorAll('script, link, style').forEach((element) => element.remove())

  const body = document.body
  while (body.firstChild) {
    body.removeChild(body.firstChild)
  }
  body.appendChild(jsContent)

  const cleanAttributes = (element) => {
    for (const child of element.children) {
      const src = child.getAttribute('src')
      const dataSrc = child.getAttribute('data-src')
      const href = child.getAttribute('href')

      while (child.attributes.length > 0) {
        child.removeAttribute(child.attributes[0].name)
      }

      if (dataSrc) {
        child.setAttribute('src', dataSrc)
      } else if (src) {
        child.setAttribute('src', src)
      }

      if (href) {
        child.setAttribute('href', href)
      }

      if (child.children.length > 0) {
        cleanAttributes(child)
      }
    }
  }

  cleanAttributes(document.body)

  const removeEmptyTags = (element) => {
    const toRemove = []

    for (const child of element.children) {
      if (child.children.length > 0) {
        removeEmptyTags(child)
      }

      const isEmpty = !child.textContent || child.textContent.trim() === ''
      const hasNoChildren = child.children.length === 0
      const isNotImg = child.tagName !== 'IMG'

      if (isEmpty && hasNoChildren && isNotImg) {
        toRemove.push(child)
      }
    }

    toRemove.forEach((elementToRemove) => elementToRemove.remove())
  }

  removeEmptyTags(document.body)

  const meta = {}
  document.querySelectorAll('meta[property^="og:"]').forEach((element) => {
    const property = element.getAttribute('property')
    const content = element.getAttribute('content')

    if (!property || !content) {
      return
    }

    const key = property.replace('og:', '').replace(/:/g, '_')
    if (key !== 'site_name') {
      meta[key] = content
    }
  })

  const contentHtml = document.body.innerHTML
  const links = []
  const contentDom = new JSDOM(contentHtml)
  contentDom.window.document.querySelectorAll('a[href]').forEach((element) => {
    try {
      const href = element.getAttribute('href')
      const url = new URL(href, 'https://mp.weixin.qq.com')
      if (url.hostname === 'mp.weixin.qq.com' && url.pathname === '/s') {
        links.push({
          title: element.textContent.trim(),
          link: url.toString()
        })
      }
    } catch {
      // ignore invalid URLs in fixture plugin
    }
  })

  meta.links = links

  const turndownService = new TurndownService()

  return {
    meta,
    contentHtml,
    markdown: turndownService.turndown(contentHtml)
  }
}

const weixinCollector = {
  name: 'weixin',
  match(url) {
    return url.hostname === 'mp.weixin.qq.com' && url.pathname === '/s'
  },
  shouldFallback(result, _ctx) {
    if (result.status !== 200) {
      return true
    }

    return parseWeixinArticle(result.body) === null
  }
}

const weixinTransformer = {
  name: 'weixin',
  normalizeDocument(ctx) {
    const html = ctx.artifacts?.rawHtml ?? ctx.response?.body ?? ''
    const article = parseWeixinArticle(html)

    if (!article) {
      return {
        url: ctx.input.url,
        title: '',
        content: {
          html: '',
          markdown: ''
        },
        assets: [],
        meta: {},
        source: 'weixin',
        tags: []
      }
    }

    return {
      url: ctx.input.url,
      title: typeof article.meta.title === 'string' ? article.meta.title : '',
      excerpt: typeof article.meta.description === 'string' ? article.meta.description : undefined,
      content: {
        html: article.contentHtml,
        markdown: article.markdown
      },
      assets: [],
      meta: {
        image: article.meta.image,
        url: article.meta.url,
        links: article.meta.links
      },
      source: 'weixin',
      tags: []
    }
  }
}

const plugin = {
  name: 'clipper-plugin-weixin',
  collectors: [weixinCollector],
  transformers: [weixinTransformer],
  publishers: []
}

export default plugin
export { plugin, weixinCollector, weixinTransformer }
